src/librustc_data_structures/stable_map.rs

```diff
@@ -0,0 +1,99 @@
+pub use rustc_hash::FxHashMap;
+use std::borrow::Borrow;
+use std::collections::hash_map::Entry;
+use std::fmt;
+use std::hash::Hash;
+
+/// A deterministic wrapper around FxHashMap that does not provide iteration support.
+///
+/// It supports insert, remove, get and get_mut functions from FxHashMap.
+/// It also allows to convert hashmap to a sorted vector with the method `into_sorted_vector()`.
+#[derive(Clone)]
+pub struct StableMap<K, V> {
+    base: FxHashMap<K, V>,
+}
+
+impl<K, V> Default for StableMap<K, V>
+where
+    K: Eq + Hash,
+{
+    fn default() -> StableMap<K, V> {
+        StableMap::new()
+    }
+}
+
+impl<K, V> fmt::Debug for StableMap<K, V>
+where
+    K: Eq + Hash + fmt::Debug,
+    V: fmt::Debug,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{:?}", self.base)
+    }
+}
+
+impl<K, V> PartialEq for StableMap<K, V>
+where
+    K: Eq + Hash,
+    V: PartialEq,
+{
+    fn eq(&self, other: &StableMap<K, V>) -> bool {
+        self.base == other.base
+    }
+}
+
+impl<K, V> Eq for StableMap<K, V>
+where
+    K: Eq + Hash,
+    V: Eq,
+{}
+
+impl<K, V> StableMap<K, V>
+where
+    K: Eq + Hash,
+{
+    pub fn new() -> StableMap<K, V> {
+        StableMap { base: FxHashMap::default() }
+    }
+
+    pub fn into_sorted_vector(self) -> Vec<(K, V)>
+    where
+        K: Ord + Copy,
+    {
+        let mut vector = self.base.into_iter().collect::<Vec<_>>();
+        vector.sort_unstable_by_key(|pair| pair.0);
```

Comment:
If you use `&pair.0` instead you wont need the `K: Copy` bound.

Reply:
Hi @bjorn3 ! Thanks for all the help so far. :)
If I use `&pair.0` or the exact statement that you had suggested, I get a lifetime error
```
error: lifetime may not live long enough
  --> src/librustc_data_structures/stable_map.rs:64:44
   |
64 |         vector.sort_unstable_by_key(|pair| &pair.0);
   |                                      ----- ^^^^^^^ returning this value requires that `'1` must outlive `'2`
   |                                      |   |
   |                                      |   return type of closure is &'2 K
   |                                      has type `&'1 (K, V)`

error: aborting due to previous error
```
Is there another workaround?

Reply:
I should have checked if it would work. I think the key would be `Copy` in most cases anyway.

Reply:
No problem. If the HashMap implementation looks fine to you, I would start on HashSet now.

Reply:
LGTM